lib/oli/delivery/attempts/activity_lifecycle/evaluate.ex

```diff
@@ -377,44 +377,81 @@ defmodule Oli.Delivery.Attempts.ActivityLifecycle.Evaluate do
-        input =
-          case p.response do
-            nil -> nil
-            map -> Map.get(map, "input")
-          end
-
-        files =
-          case p.response do
-            nil -> nil
-            map -> Map.get(map, "files", [])
+  def update_part_attempts_and_get_activity_attempts(resource_attempt_id, datashop_session_id) do
+    resource_attempt_id
+    |> get_latest_activity_attempts()
+    |> Enum.reduce_while(
+      {0, [], [], []},
+      fn activity_attempt,
+         {
+           i,
+           activity_attempt_values,
+           activity_attempt_params,
+           part_attempt_guids
+         } = acc ->
+        if activity_attempt.lifecycle_state != :evaluated and activity_attempt.scoreable do
+          # --- part attempts
+          activity_attempt = activity_attempt |> Repo.preload([:revision, :resource_attempt])
+          part_attempts = get_latest_part_attempts(activity_attempt.attempt_guid)
+
+          part_inputs =
+            part_attempts
+            |> Enum.map(fn pa ->
+              {input, files} =
+                if pa.response,
+                  do: {Map.get(pa.response, "input"), Map.get(pa.response, "files", [])},
+                  else: {nil, nil}
+
+              %{
+                attempt_guid: pa.attempt_guid,
+                input: %StudentInput{input: input, files: files}
+              }
+            end)
+            |> filter_already_evaluated(part_attempts)
+
+          case activity_attempt
+               |> do_evaluate_submissions(part_inputs, part_attempts)
+               |> persist_evaluations(part_inputs, fn result -> result end, datashop_session_id) do
+            {:ok, _} ->
+              # --- activity attempt
+              part_attempts = get_latest_part_attempts(activity_attempt.attempt_guid)
+
+              %{
+                score: score,
+                out_of: out_of,
+                lifecycle_state: lifecycle_state,
+                date_evaluated: date_evaluated,
+                date_submitted: date_submitted
+              } = determine_activity_rollup_attrs(part_inputs, part_attempts, activity_attempt)
+
+              # ---
+              {:cont,
+               {
+                 i + 6,
+                 activity_attempt_values ++
+                   [
+                     "($#{i + 1}, $#{i + 2}::double precision, $#{i + 3}::double precision, $#{i + 4}, $#{i + 5}::timestamp, $#{i + 6}::timestamp)"
+                   ],
+                 activity_attempt_params ++
+                   [
+                     activity_attempt.attempt_guid,
+                     score,
+                     out_of,
+                     Atom.to_string(lifecycle_state),
+                     date_evaluated,
+                     date_submitted
+                   ],
+                 part_attempt_guids ++
+                   Enum.map(part_attempts, fn part_attempt -> part_attempt.attempt_guid end)
+               }}
+
+            error -> {:halt, error}
```

Comment:
We're doing a `reduce_while` so we can stop the iteration returning a `halt`. 

Did that because if the part attempts fail to update and we continue, there will be invalid data to update the activity attempt (and then will be invalid data for the resource attempt, etc etc), so I'd rather stop the iteration and since everything is inside a transaction, roll back all the changes.

Anyway I'm really open to any other suggestions here

Reply:
Excellent. Very clear explanation. Thank you for it. 🙌